library/psa_crypto.c

```diff
@@ -6684,36 +6684,32 @@ static psa_status_t psa_pbkdf2_set_salt(psa_pbkdf2_key_derivation_t *pbkdf2,
                                         const uint8_t *data,
                                         size_t data_length)
 {
-    if (pbkdf2->state != PSA_PBKDF2_STATE_INPUT_COST_SET &&
-        pbkdf2->state != PSA_PBKDF2_STATE_SALT_SET) {
+    if (pbkdf2->state == PSA_PBKDF2_STATE_INPUT_COST_SET) {
+        pbkdf2->state = PSA_PBKDF2_STATE_SALT_SET;
```

Comment:
Post-refactoring, the initialisation of the salt length at the start of a salt is now missing for all salt lengths. I think a zero-initialisation here is needed:
```suggestion
        pbkdf2->state = PSA_PBKDF2_STATE_SALT_SET;
        pbkdf2->salt_length = 0;
```

Reply:
If the salt hasn't been set yet then `pbkdf2->salt == NULL` and `pbkdf2->salt_length == 0`. The former is necessary for `abort`, which calls `free(salt)` without caring about the state. The latter is necessary to preserve the invariant that `{salt, salt_length}` is a valid buffer, which I'm not sure is otherwise strictly necessary but is generally a good idea. Since there are strong reasons why `salt_length` must be 0, I don't see a strong need for the redundant assignment.

In high-level terms, the salt starts out empty, and inputs can append to it. This is orthogonal to the state machine which avoids some user errors and enforces application portability by requiring the inputs to be presented in a certain order. This orthogonality is a weak argument against handling `salt` and `salt_length` differently based on the state. Code size is another argument.

So on balance, I maintain that the assignment is better left out.